File: Programmers/Algorithm_Kit/02_Stack_Queue/002.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    vector<int> answer;
    vector<int> days;

    for (int i = 0; i < progresses.size(); i++) {
        int remain = 100 - progresses[i];
        int day = remain / speeds[i];

        if (remain % speeds[i]) {
            day++;
        }

        days.push_back(day);
    }

    int deploy_day = days[0];
    int count = 1;
    
    for (int i = 1; i < days.size(); i++) {
        if (days[i] <= deploy_day) {
            count++;
        }
        else {
            answer.push_back(count);

            deploy_day = days[i];
            count = 1;
        }
    }

    answer.push_back(count);

    return answer;
}

File: programmers/algorithmkit/02_stack_queue/002.java
import java.util.*;

class Solution {
    public List<Integer> solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
        List<Integer> days = new ArrayList<>();

        for (int i = 0; i < progresses.length; i++) {
            int remaining_work = 100 - progresses[i];
            int remaining_days = remaining_work / speeds[i];

            if (remaining_work % speeds[i] != 0) {
                remaining_days++;
            }

            days.add(remaining_days);
        }

        // 첫 원소부터 반복문에 넣어야 한다는 강박 버리기
        int deploy_day = days.get(0);
        int cnt = 1;

        for (int i = 1; i < days.size(); i++) {
            int temp = days.get(i);

            if (deploy_day < temp) {
                answer.add(cnt);

                deploy_day = temp;
                cnt = 1;
            } else {
                cnt++;
            }
        }

        answer.add(cnt);

        return answer;
    }
}
